src/tree/param.h

```diff
@@ -271,54 +271,61 @@ XGBOOST_DEVICE inline static T Sqr(T a) { return a * a; }
 
 // calculate the cost of loss function
 template <typename TrainingParams, typename T>
-XGBOOST_DEVICE inline T CalcGainGivenWeight(const TrainingParams &p, T sum_grad,
-                                        T sum_hess, T w) {
-  return -(2.0 * sum_grad * w + (sum_hess + p.reg_lambda) * Sqr(w));
+XGBOOST_DEVICE inline T CalcGainGivenWeight(const TrainingParams &p,
+                                            T sum_grad, T sum_hess, T w) {
+  return -(T(2.0) * sum_grad * w + (sum_hess + p.reg_lambda) * Sqr(w));
 }
 
 // calculate the cost of loss function
 template <typename TrainingParams, typename T>
 XGBOOST_DEVICE inline T CalcGain(const TrainingParams &p, T sum_grad, T sum_hess) {
   if (sum_hess < p.min_child_weight) {
     return T(0.0);
-}
+  }
   if (p.max_delta_step == 0.0f) {
     if (p.reg_alpha == 0.0f) {
       return Sqr(sum_grad) / (sum_hess + p.reg_lambda);
     } else {
       return Sqr(ThresholdL1(sum_grad, p.reg_alpha)) /
-             (sum_hess + p.reg_lambda);
+          (sum_hess + p.reg_lambda);
     }
   } else {
     T w = CalcWeight(p, sum_grad, sum_hess);
-    T ret = sum_grad * w + T(0.5) * (sum_hess + p.reg_lambda) * Sqr(w);
+    T ret = CalcGainGivenWeight(p, sum_grad, sum_hess, w);
```

Comment:
Are you absolutely sure this is equivalent? I would like a little more assurance here because of this function is wrong we would have a big problem.

Reply:
Thanks. Note that due to the non associativity of floating point operations the result may be very slightly different, as a consequence from this commit onwards xgboost may produce very slightly different results.

I think we should still merge this change, if we enforced ourselves to produce identical xgboost results in every release we would never have a chance to improve xgboost.